File: 010-Regular Expression Matching.c
#include<stdio.h>
#include<stdbool.h>
bool isMatch(char* s, char* p) {
    if((p==NULL)||(s==NULL))
        return false;
    //如果p匹配到最后是空，那么如果s也是空，就返回true，否则就是返回false
    if(!*p) return !*s; 
    if(*(p+1)== '*'){
        while((*p == *s)||(*s && *p == '.')){
            if (isMatch(s,p+2))return true;
            s++;
        }
        return isMatch(s,p+2);
    }else if((*p==*s)||(*s && *p=='.')){
        printf("%c\n",*(s+1));
        printf("%c\n",*(p+1));
        return isMatch(s+1,p+1);
    }
    return false;
    
}
int main(){
    isMatch("aa","aa");
    return 0;
}

File: 010-Regular Expression Matching.py
class Solution(object):
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        #
        if len(s)<1 or len(p)<1:
            return False
        if s[0] == p[0] or p[0] == '.':
                if len(p) >= 2:
                    if p[1] != '*':
                        return self.isMatch(s[1:],p[1:])
                    #p[1]=='*'
                    if len(p) == 2:
                        if p[0] == '.':
                            return True
                        else:
                            for i in s:
                                if i != p[0]:
                                    return False
                            return True
                    else:
                        if p[0] == '.':
                            i=0
                            while(i<len(s)):
                                if self.isMatch(s[i:],p[2:]):
                                    return True
                                i += 1
                            return False
                        else:
                            index=0
                            while(index < len(s)):
                                if s[index]==p[0]:
                                    index += 1
                            a=0
                            while(a <= index):
                                if self.isMatch(s[a:],p[2:]):
                                    return True
                                a += 1
                            return False 
                        
           
                elif len(s) == 1:#len(s)=len(p)=1
                    return True
                else:#len(s)>len(p)=1
                    return False
           
        elif s[0] != p[0]:
            if len(p) >= 2:
                if p[1] == '*':
                    return self.isMatch(s,p[2:])
            return False   
        else:
                return False
if __name__ == '__main__':
    s=Solution()
    print(s.isMatch("aab","c*a*b"))